File: mobile/android/src/main/java/com/yuledkcs/KeyManager.kt
package com.yuledkcs

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.yuledkcs.model.DigitalKey
import com.yuledkcs.model.KeyPermission
import com.yuledkcs.model.KeyUsageLog
import com.yuledkcs.model.Permission
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.Date
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Manager for digital key operations
 * 
 * Handles issuing, listing, sharing, and revoking digital keys.
 * Also manages key usage logs and offline operations.
 */
class KeyManager private constructor(private val context: Context) {
    
    companion object {
        private const val TAG = "KeyManager"
        private const val PREFS_NAME = "key_manager_prefs"
        private const val KEY_LOGS_PREFIX = "usage_logs_"
        
        @Volatile
        private var instance: KeyManager? = null
        
        fun getInstance(context: Context): KeyManager {
            return instance ?: synchronized(this) {
                instance ?: KeyManager(context.applicationContext).also { instance = it }
            }
        }
    }
    
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val keysCache = ConcurrentHashMap<String, DigitalKey>()
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    
    // StateFlow for reactive UI updates
    private val _keysFlow = MutableStateFlow<List<DigitalKey>>(emptyList())
    val keysFlow: StateFlow<List<DigitalKey>> = _keysFlow
    
    private val _activeKeyFlow = MutableStateFlow<DigitalKey?>(null)
    val activeKeyFlow: StateFlow<DigitalKey?> = _activeKeyFlow

    init {
        // Load cached keys on initialization
        scope.launch {
            loadCachedKeys()
        }
    }
    
    /**
     * Issue a new digital key for a vehicle
     * 
     * @param vehicleId The vehicle identifier
     * @param callback Callback with Result containing the issued DigitalKey
     */
    fun issueKey(
        vehicleId: String,
        callback: (Result<DigitalKey>) -> Unit
    ) {
        YuleDKCS.checkInitialized()
        
        scope.launch {
            try {
                val response = YuleDKCS.api.issueKey(
                    apiKey = YuleDKCS.apiKey,
                    request = IssueKeyRequest(vehicleId)
                )
                
                if (response.isSuccessful) {
                    val key = response.body()
                    if (key != null) {
                        // Store key securely
                        storeKey(key)
                        keysCache[key.id] = key
                        _keysFlow.value = keysCache.values.toList()
                        
                        withContext(Dispatchers.Main) {
                            callback(Result.success(key))
                        }
                    } else {
                        throw IllegalStateException("Empty response body")
                    }
                } else {
                    throw Exception("Failed to issue key: ${response.code()}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error issuing key", e)
                withContext(Dispatchers.Main) {
                    callback(Result.failure(e))
                }
            }
        }
    }
    
    /**
     * Receive a shared key via QR code or deep link
     * 
     * @param shareToken The share token from QR code or link
     * @param callback Callback with Result containing the received DigitalKey
     */
    fun receiveSharedKey(
        shareToken: String,
        callback: (Result<DigitalKey>) -> Unit
    ) {
        YuleDKCS.checkInitialized()
        
        scope.launch {
            try {
                val response = YuleDKCS.api.receiveSharedKey(
                    apiKey = YuleDKCS.apiKey,
                    token = shareToken
                )
                
                if (response.isSuccessful) {
                    val key = response.body()
                    if (key != null) {
                        storeKey(key)
                        keysCache[key.id] = key
                        _keysFlow.value = keysCache.values.toList()
                        
                        withContext(Dispatchers.Main) {
                            callback(Result.success(key))
                        }
                    } else {
                        throw IllegalStateException("Empty response body")
                    }
                } else {
                    throw Exception("Failed to receive key: ${response.code()}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error receiving shared key", e)
                withContext(Dispatchers.Main) {
                    callback(Result.failure(e))
                }
            }
        }
    }
    
    /**
     * Activate a key (bind to SE)
     * 
     * @param keyId The key identifier
     * @param callback Callback with Result
     */
    fun activateKey(
        keyId: String,
        callback: ((Result<Unit>) -> Unit)? = null
    ) {
        YuleDKCS.checkInitialized()
        
        scope.launch {
            try {
                // First activate on server
                val response = YuleDKCS.api.activateKey(
                    apiKey = YuleDKCS.apiKey,
                    keyId = keyId
                )
                
                if (response.isSuccessful) {
                    // Then store in SE
                    val key = keysCache[keyId]
                    if (key != null) {
                        YuleDKCS.securityProvider.storeKeyInSE(keyId, key.keyData)
                        
                        // Update local status
                        val updatedKey = key.copy(status = "active")
                        keysCache[keyId] = updatedKey
                        _keysFlow.value = keysCache.values.toList()
                        
                        // Update active key
                        _activeKeyFlow.value = updatedKey
                    }
                    
                    withContext(Dispatchers.Main) {
                        callback?.invoke(Result.success(Unit))
                    }
                } else {
                    throw Exception("Failed to activate key: ${response.code()}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error activating key", e)
                withContext(Dispatchers.Main) {
                    callback?.invoke(Result.failure(e))
                }
            }
        }
    }
    
    /**
     * Get list of all keys
     * 
     * @param forceRefresh Force refresh from server
     * @return List of DigitalKey objects
     */
    fun listKeys(forceRefresh: Boolean = false): List<DigitalKey> {
        YuleDKCS.checkInitialized()
        
        if (forceRefresh || keysCache.isEmpty()) {
            refreshKeys()
        }
        
        return keysCache.values.toList()
    }
    
    /**
     * Get a specific key by ID
     * 
     * @param keyId The key identifier
     * @return DigitalKey or null if not found
     */
    fun getKey(keyId: String): DigitalKey? {
        return keysCache[keyId] ?: loadStoredKey(keyId)
    }
    
    /**
     * Set active key for quick access
     * 
     * @param keyId The key identifier
     */
    fun setActiveKey(keyId: String?) {
        _activeKeyFlow.value = keyId?.let { keysCache[it] }
        prefs.edit().putString("active_key_id", keyId).apply()
    }
    
    /**
     * Get the currently active key
     */
    fun getActiveKey(): DigitalKey? {
        val activeId = prefs.getString("active_key_id", null)
        return activeId?.let { getKey(it) }
    }
    
    /**
     * Share a key with another user
     * 
     * @param keyId The key identifier to share
     * @param to Email or phone number of recipient
     * @param permissions List of permissions to grant
     * @param expiresInDays Expiration in days
     * @param callback Optional callback for result
     */
    fun shareKey(
        keyId: String,
        to: String,
        permissions: List<Permission>,
        expiresInDays: Int = 7,
        callback: ((Result<String>) -> Unit)? = null
    ) {
        YuleDKCS.checkInitialized()
        
        scope.launch {
            try {
                val response = YuleDKCS.api.shareKey(
                    apiKey = YuleDKCS.apiKey,
                    keyId = keyId,
                    request = ShareKeyRequest(
                        recipient = to,
                        permissions = permissions.map { it.name },
                        expiresInDays = expiresInDays
                    )
                )
                
                withContext(Dispatchers.Main) {
                    if (response.isSuccessful) {
                        val shareResult = response.body()
                        callback?.invoke(Result.success(shareResult?.shareLink ?: ""))
                    } else {
                        callback?.invoke(
                            Result.failure(
                                Exception("Failed to share key: ${response.code()}")
                            )
                        )
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error sharing key", e)
                withContext(Dispatchers.Main) {
                    callback?.invoke(Result.failure(e))
                }
            }
        }
    }
    
    /**
     * Revoke a key
     * 
     * @param keyId The key identifier to revoke
     * @param callback Optional callback for result
     */
    fun revokeKey(
        keyId: String,
        callback: ((Result<Unit>) -> Unit)? = null
    ) {
        YuleDKCS.checkInitialized()
        
        scope.launch {
            try {
                val response = YuleDKCS.api.revokeKey(
                    apiKey = YuleDKCS.apiKey,
                    keyId = keyId
                )
                
                if (response.isSuccessful) {
                    keysCache.remove(keyId)
                    deleteStoredKey(keyId)
                    _keysFlow.value = keysCache.values.toList()
                    
                    // Clear active key if it was the revoked one
                    if (_activeKeyFlow.value?.id == keyId) {
                        _activeKeyFlow.value = null
                        prefs.edit().remove("active_key_id").apply()
                    }
                }
                
                withContext(Dispatchers.Main) {
                    if (response.isSuccessful) {
                        callback?.invoke(Result.success(Unit))
                    } else {
                        callback?.invoke(
                            Result.failure(
                                Exception("Failed to revoke key: ${response.code()}")
                            )
                        )
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error revoking key", e)
                withContext(Dispatchers.Main) {
                    callback?.invoke(Result.failure(e))
                }
            }
        }
    }
    
    /**
     * Delete local key data only (keep server key)
     * Use this when user wants to remove key from this device only
     * 
     * @param keyId The key identifier
     */
    fun deleteLocalKey(keyId: String) {
        keysCache.remove(keyId)
        deleteStoredKey(keyId)
        _keysFlow.value = keysCache.values.toList()
        
        if (_activeKeyFlow.value?.id == keyId) {
            _activeKeyFlow.value = null
            prefs.edit().remove("active_key_id").apply()
        }
    }
    
    /**
     * Record a key usage log
     * 
     * @param keyId The key identifier
     * @param operation The operation performed
     * @param status Success or failure status
     * @param location Optional location
     */
    fun recordUsage(
        keyId: String,
        operation: String,
        status: KeyUsageLog.Status,
        location: Pair<Double, Double>? = null,
        failureReason: String? = null
    ) {
        val log = KeyUsageLog(
            id = UUID.randomUUID().toString(),
            keyId = keyId,
            operation = operation,
            status = status,
            timestamp = Date(),
            location = location?.let { KeyUsageLog.Location(it.first, it.second) },
            deviceInfo = android.os.Build.MODEL,
            failureReason = failureReason
        )
        
        // Store locally
        storeUsageLog(keyId, log)
        
        // Try to sync to server
        scope.launch {
            try {
                YuleDKCS.api.recordUsage(
                    apiKey = YuleDKCS.apiKey,
                    keyId = keyId,
                    log = log
                )
            } catch (e: Exception) {
                Log.w(TAG, "Failed to sync usage log, will retry later", e)
            }
        }
    }
    
    /**
     * Get usage logs for a key
     * 
     * @param keyId The key identifier
     * @return List of usage logs
     */
    fun getUsageLogs(keyId: String): List<KeyUsageLog> {
        return loadUsageLogs(keyId)
    }
    
    /**
     * Check if a key has a specific permission
     * 
     * @param keyId The key identifier
     * @param permission The permission to check
     * @return true if key has the permission
     */
    fun hasPermission(keyId: String, permission: Permission): Boolean {
        val key = keysCache[keyId] ?: return false
        return key.permissions.any { it.type == permission.name && it.enabled }
    }
    
    /**
     * Check if key is expired
     * 
     * @param keyId The key identifier
     * @return true if key is expired
     */
    fun isKeyExpired(keyId: String): Boolean {
        val key = keysCache[keyId] ?: return true
        key.expiresAt?.let {
            return Date().after(it)
        }
        return false
    }
    
    // Private helper methods
    
    private fun loadCachedKeys() {
        // Load from secure storage
        val keyIds = prefs.getStringSet("key_ids", emptySet()) ?: emptySet()
        keyIds.forEach { keyId ->
            loadStoredKey(keyId)?.let { keysCache[keyId] = it }
        }
        _keysFlow.value = keysCache.values.toList()
        
        // Load active key
        val activeId = prefs.getString("active_key_id", null)
        _activeKeyFlow.value = activeId?.let { keysCache[it] }
        
        // Refresh from server
        refreshKeys()
    }
    
    private fun refreshKeys() {
        scope.launch {
            try {
                val response = YuleDKCS.api.listKeys(YuleDKCS.apiKey)
                if (response.isSuccessful) {
                    response.body()?.let { keys ->
                        keysCache.clear()
                        keys.forEach { 
                            keysCache[it.id] = it
                            storeKey(it)
                        }
                        _keysFlow.value = keysCache.values.toList()
                        
                        // Update key ID set
                        prefs.edit().putStringSet("key_ids", keysCache.keys).apply()
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error refreshing keys", e)
            }
        }
    }
    
    private fun storeKey(key: DigitalKey) {
        YuleDKCS.securityProvider.storeSecureData(
            "key_${key.id}",
            key.keyData
        )
    }
    
    private fun loadStoredKey(keyId: String): DigitalKey? {
        val keyData = YuleDKCS.securityProvider.getSecureData("key_$keyId") ?: return null
        return DigitalKey(
            id = keyId,
            keyData = keyData
        )
    }
    
    private fun deleteStoredKey(keyId: String) {
        YuleDKCS.securityProvider.deleteSecureData("key_$keyId")
    }
    
    private fun storeUsageLog(keyId: String, log: KeyUsageLog) {
        val logs = loadUsageLogs(keyId).toMutableList()
        logs.add(0, log) // Add to beginning
        
        // Keep only last 100 logs
        if (logs.size > 100) {
            logs.subList(100, logs.size).clear()
        }
        
        // Serialize and store
        val json = logs.joinToString("|") { it.toJson() }
        prefs.edit().putString("$KEY_LOGS_PREFIX$keyId", json).apply()
    }
    
    private fun loadUsageLogs(keyId: String): List<KeyUsageLog> {
        val json = prefs.getString("$KEY_LOGS_PREFIX$keyId", null) ?: return emptyList()
        return json.split("|").mapNotNull { KeyUsageLog.fromJson(it) }
    }
}

// API Request/Response data classes
internal data class IssueKeyRequest(val vehicleId: String)
internal data class ShareKeyRequest(
    val recipient: String,
    val permissions: List<String>,
    val expiresInDays: Int = 7
)

File: mobile/ios/Sources/YuleDKCS/KeyManager.swift
import Foundation
import Combine
import CoreBluetooth
import LocalAuthentication
import CryptoKit

/// Manager for digital key operations
///
/// Handles issuing, listing, sharing, and revoking digital keys.
/// Also manages key usage logs and offline operations.
public class KeyManager: ObservableObject {
    
    public static let shared = KeyManager()
    
    @Published public private(set) var keys: [DigitalKey] = []
    @Published public private(set) var activeKey: DigitalKey?
    @Published public private(set) var isLoading = false
    
    private var cancellables = Set<AnyCancellable>()
    private let userDefaults = UserDefaults.standard
    private let keychain = KeychainManager.shared
    private let queue = DispatchQueue(label: "com.yuledkcs.keymanager", qos: .userInitiated)
    
    private init() {
        loadCachedKeys()
    }
    
    // MARK: - Key Lifecycle
    
    /// Issue a new digital key for a vehicle
    /// - Parameters:
    ///   - vehicleId: The vehicle identifier
    ///   - completion: Callback with Result containing the issued DigitalKey
    public func issueKey(
        vehicleId: String,
        completion: @escaping (Result<DigitalKey, Error>) -> Void
    ) {
        guard YuleDKCS.shared.isInitialized else {
            completion(.failure(YuleDKCSError.notInitialized))
            return
        }
        
        isLoading = true
        
        YuleDKCS.shared.api.issueKey(vehicleId: vehicleId)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] result in
                    self?.isLoading = false
                    if case .failure(let error) = result {
                        completion(.failure(error))
                    }
                },
                receiveValue: { [weak self] key in
                    self?.storeKey(key)
                    completion(.success(key))
                }
            )
            .store(in: &cancellables)
    }
    
    /// Receive a shared key via QR code or deep link
    /// - Parameters:
    ///   - shareToken: The share token from QR code or link
    ///   - completion: Callback with Result containing the received DigitalKey
    public func receiveSharedKey(
        shareToken: String,
        completion: @escaping (Result<DigitalKey, Error>) -> Void
    ) {
        guard YuleDKCS.shared.isInitialized else {
            completion(.failure(YuleDKCSError.notInitialized))
            return
        }
        
        isLoading = true
        
        YuleDKCS.shared.api.receiveSharedKey(token: shareToken)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] result in
                    self?.isLoading = false
                    if case .failure(let error) = result {
                        completion(.failure(error))
                    }
                },
                receiveValue: { [weak self] key in
                    self?.storeKey(key)
                    completion(.success(key))
                }
            )
            .store(in: &cancellables)
    }
    
    /// Activate a key (bind to Secure Enclave)
    /// - Parameters:
    ///   - keyId: The key identifier
    ///   - completion: Callback with Result
    public func activateKey(
        keyId: String,
        completion: ((Result<Void, Error>) -> Void)? = nil
    ) {
        guard YuleDKCS.shared.isInitialized else {
            completion?(.failure(YuleDKCSError.notInitialized))
            return
        }
        
        queue.async { [weak self] in
            // First activate on server
            YuleDKCS.shared.api.activateKey(keyId: keyId)
                .receive(on: DispatchQueue.main)
                .sink(
                    receiveCompletion: { result in
                        if case .failure(let error) = result {
                            completion?(.failure(error))
                        }
                    },
                    receiveValue: { [weak self] _ in
                        guard let self = self else { return }
                        
                        // Store in Secure Enclave
                        if let key = self.keys.first(where: { $0.id == keyId }) {
                            do {
                                try self.keychain.storeKey(keyId: keyId, keyData: key.keyData)
                                
                                // Update local status
                                var updatedKey = key
                                updatedKey.status = .active
                                self.updateKey(updatedKey)
                                self.activeKey = updatedKey
                                
                                completion?(.success(()))
                            } catch {
                                completion?(.failure(error))
                            }
                        }
                    }
                )
                .store(in: &self!.cancellables)
        }
    }
    
    /// Get list of all keys
    /// - Parameter forceRefresh: Force refresh from server
    /// - Returns: Array of DigitalKey
    public func listKeys(forceRefresh: Bool = false) -> [DigitalKey] {
        guard YuleDKCS.shared.isInitialized else {
            return []
        }
        
        if forceRefresh || keys.isEmpty {
            refreshKeys()
        }
        
        return keys
    }
    
    /// Get a specific key by ID
    /// - Parameter keyId: The key identifier
    /// - Returns: DigitalKey or nil
    public func getKey(keyId: String) -> DigitalKey? {
        return keys.first { $0.id == keyId }
    }
    
    /// Set active key for quick access
    /// - Parameter keyId: The key identifier
    public func setActiveKey(keyId: String?) {
        activeKey = keyId.flatMap { getKey(keyId: $0) }
        userDefaults.set(keyId, forKey: "active_key_id")
    }
    
    /// Get the currently active key
    public func getActiveKey() -> DigitalKey? {
        return activeKey
    }
    
    // MARK: - Sharing
    
    /// Share a key with another user
    /// - Parameters:
    ///   - keyId: The key identifier to share
    ///   - recipient: Email or phone number of recipient
    ///   - permissions: List of permissions to grant
    ///   - expiresInDays: Expiration in days
    ///   - completion: Callback with Result containing share link
    public func shareKey(
        keyId: String,
        recipient: String,
        permissions: [Permission],
        expiresInDays: Int = 7,
        completion: ((Result<String, Error>) -> Void)? = nil
    ) {
        guard YuleDKCS.shared.isInitialized else {
            completion?(.failure(YuleDKCSError.notInitialized))
            return
        }
        
        YuleDKCS.shared.api.shareKey(
            keyId: keyId,
            recipient: recipient,
            permissions: permissions.map { $0.rawValue },
            expiresInDays: expiresInDays
        )
        .receive(on: DispatchQueue.main)
        .sink(
            receiveCompletion: { result in
                if case .failure(let error) = result {
                    completion?(.failure(error))
                }
            },
            receiveValue: { response in
                completion?(.success(response.shareLink))
            }
        )
        .store(in: &cancellables)
    }
    
    // MARK: - Revocation
    
    /// Revoke a key
    /// - Parameters:
    ///   - keyId: The key identifier to revoke
    ///   - completion: Callback with Result
    public func revokeKey(
        keyId: String,
        completion: ((Result<Void, Error>) -> Void)? = nil
    ) {
        guard YuleDKCS.shared.isInitialized else {
            completion?(.failure(YuleDKCSError.notInitialized))
            return
        }
        
        YuleDKCS.shared.api.revokeKey(keyId: keyId)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { result in
                    if case .failure(let error) = result {
                        completion?(.failure(error))
                    }
                },
                receiveValue: { [weak self] _ in
                    self?.removeKey(keyId)
                    completion?(.success(()))
                }
            )
            .store(in: &cancellables)
    }
    
    /// Delete local key data only (keep server key)
    public func deleteLocalKey(keyId: String) {
        removeKey(keyId)
    }
    
    // MARK: - Usage Logging
    
    /// Record a key usage log
    public func recordUsage(
        keyId: String,
        operation: String,
        status: KeyUsageLog.Status,
        location: CLLocationCoordinate2D? = nil,
        failureReason: String? = nil
    ) {
        let log = KeyUsageLog(
            id: UUID().uuidString,
            keyId: keyId,
            operation: operation,
            status: status,
            timestamp: Date(),
            location: location.flatMap { KeyUsageLog.Location(lat: $0.latitude, lng: $0.longitude) },
            deviceInfo: UIDevice.current.model,
            failureReason: failureReason
        )
        
        // Store locally
        storeUsageLog(keyId: keyId, log: log)
        
        // Try to sync to server
        YuleDKCS.shared.api.recordUsage(keyId: keyId, log: log)
            .sink(receiveCompletion: { _ in }, receiveValue: { _ in })
            .store(in: &cancellables)
    }
    
    /// Get usage logs for a key
    public func getUsageLogs(keyId: String) -> [KeyUsageLog] {
        return loadUsageLogs(keyId: keyId)
    }
    
    // MARK: - Permission Checking
    
    /// Check if a key has a specific permission
    public func hasPermission(keyId: String, permission: Permission) -> Bool {
        guard let key = getKey(keyId: keyId) else { return false }
        return key.permissions.contains { $0.type == permission.rawValue && $0.enabled }
    }
    
    /// Check if key is expired
    public func isKeyExpired(keyId: String) -> Bool {
        guard let key = getKey(keyId: keyId) else { return true }
        guard let expiresAt = key.expiresAt else { return false }
        return Date() > expiresAt
    }
    
    // MARK: - Private Methods
    
    private func loadCachedKeys() {
        if let data = userDefaults.data(forKey: "cached_keys"),
           let keys = try? JSONDecoder().decode([DigitalKey].self, from: data) {
            self.keys = keys
        }
        
        if let activeId = userDefaults.string(forKey: "active_key_id") {
            activeKey = getKey(keyId: activeId)
        }
        
        refreshKeys()
    }
    
    private func refreshKeys() {
        YuleDKCS.shared.api.listKeys()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { _ in },
                receiveValue: { [weak self] keys in
                    self?.keys = keys
                    self?.saveCachedKeys()
                }
            )
            .store(in: &cancellables)
    }
    
    private func storeKey(_ key: DigitalKey) {
        if !keys.contains(where: { $0.id == key.id }) {
            keys.append(key)
            saveCachedKeys()
            
            // Store key data securely
            try? keychain.storeKey(keyId: key.id, keyData: key.keyData)
        }
    }
    
    private func updateKey(_ key: DigitalKey) {
        if let index = keys.firstIndex(where: { $0.id == key.id }) {
            keys[index] = key
            saveCachedKeys()
        }
    }
    
    private func removeKey(_ keyId: String) {
        keys.removeAll { $0.id == keyId }
        saveCachedKeys()
        
        try? keychain.deleteKey(keyId: keyId)
        
        if activeKey?.id == keyId {
            activeKey = nil
            userDefaults.removeObject(forKey: "active_key_id")
        }
    }
    
    private func saveCachedKeys() {
        if let data = try? JSONEncoder().encode(keys) {
            userDefaults.set(data, forKey: "cached_keys")
        }
    }
    
    private func storeUsageLog(keyId: String, log: KeyUsageLog) {
        var logs = loadUsageLogs(keyId: keyId)
        logs.insert(log, at: 0)
        
        // Keep only last 100 logs
        if logs.count > 100 {
            logs = Array(logs.prefix(100))
        }
        
        if let data = try? JSONEncoder().encode(logs) {
            userDefaults.set(data, forKey: "usage_logs_\(keyId)")
        }
    }
    
    private func loadUsageLogs(keyId: String) -> [KeyUsageLog] {
        guard let data = userDefaults.data(forKey: "usage_logs_\(keyId)") else {
            return []
        }
        return (try? JSONDecoder().decode([KeyUsageLog].self, from: data)) ?? []
    }
}

// MARK: - Errors

public enum YuleDKCSError: Error {
    case notInitialized
    case keyNotFound
    case permissionDenied
    case bleNotAvailable
    case connectionFailed
    case commandFailed
}
